File: test/test_hmm.cc
#include <iostream>

#include <cassert>

#include "pml_hmm.hpp"

using namespace pml;

int main(){

  size_t N = 3;
  size_t T = 100;

  Vector p1 = normalize(Vector::ones(N));
  Matrix A = Matrix::identity(N) * 0.9;
  A(1,0) = 0.1;
  A(2,1) = 0.1;
  A(0,2) = 0.1;
  Matrix B = normalize(Matrix::identity(N) * 0.8 + Uniform().rand(N,N) * 0.2);


  DiscreteHMM dhmm(p1, A, B);

  Vector states, obs;
  std::tie(states, obs) = dhmm.generateData(T);

  auto alpha = dhmm.forward(obs);
  auto beta = dhmm.backward(obs);
  Matrix gamma = alpha + beta;



  dhmm.save("/tmp");
  states.saveTxt("/tmp/states.txt");
  obs.saveTxt("/tmp/obs.txt");
  normalizeExp(alpha, 0).saveTxt("/tmp/alpha.txt");
  normalizeExp(beta, 0).saveTxt("/tmp/beta.txt");
  normalizeExp(gamma, 0).saveTxt("/tmp/gamma.txt");

  return 0;
}

File: test/python/test_hmm.py
import pml
import matplotlib.pyplot as plt


def plot_dhmm():

    states = pml.loadTxt("/tmp/states.txt")
    obs = pml.loadTxt("/tmp/obs.txt")

    alpha = pml.loadTxt("/tmp/alpha.txt")
    beta = pml.loadTxt("/tmp/beta.txt")
    gamma = pml.loadTxt("/tmp/gamma.txt")

    fig, ax = plt.subplots(3, sharex=True, figsize=(18,10))

    ax[0].pcolormesh(alpha, cmap=plt.cm.Greys, vmin=0, vmax=1)
    ax[0].plot(range(len(states)), states + 0.5, 'r*')
    ax[0].plot(range(len(obs)), obs + 0.4, 'b*')
    ax[0].set_title("Forward Recursion")

    ax[1].pcolormesh(beta, cmap=plt.cm.Greys, vmin=0, vmax=1)
    ax[1].plot(range(len(states)), states + 0.5, 'r*')
    ax[1].plot(range(len(obs)), obs + 0.4, 'b*')
    ax[1].set_title("Backward Recursion")

    ax[2].pcolormesh(gamma, cmap=plt.cm.Greys, vmin=0, vmax=1)
    ax[2].plot(range(len(states)), states + 0.5, 'r*')
    ax[2].plot(range(len(obs)), obs + 0.4, 'b*')
    ax[2].set_title("Smoothed")
    ax[2].set_xlabel("Time")

    plt.show()

if __name__ == '__main__':
    plot_dhmm()